torchrl/data/replay_buffers/samplers.py

```diff
@@ -0,0 +1,179 @@
+        "Schaul, T.; Quan, J.; Antonoglou, I.; and Silver, D. 2015.
+        Prioritized experience replay."
+        (https://arxiv.org/abs/1511.05952)
+
+    Args:
+        alpha (float): exponent α determines how much prioritization is used,
+            with α = 0 corresponding to the uniform case.
+        beta (float): importance sampling negative exponent.
+        eps (float): delta added to the priorities to ensure that the buffer
+            does not contain null priorities.
+    """
+
+    def __init__(
+        self,
+        max_capacity: int,
+        alpha: float,
+        beta: float,
+        eps: float = 1e-8,
+        dtype: torch.dtype = torch.float,
+    ) -> None:
+        if alpha <= 0:
+            raise ValueError(
+                f"alpha must be strictly greater than 0, got alpha={alpha}"
+            )
+        if beta < 0:
+            raise ValueError(f"beta must be greater or equal to 0, got beta={beta}")
+
+        self._max_capacity = max_capacity
+        self._alpha = alpha
+        self._beta = beta
+        self._eps = eps
+        if dtype in (torch.float, torch.FloatType, torch.float32):
+            self._sum_tree = SumSegmentTreeFp32(self._max_capacity)
+            self._min_tree = MinSegmentTreeFp32(self._max_capacity)
+        elif dtype in (torch.double, torch.DoubleTensor, torch.float64):
+            self._sum_tree = SumSegmentTreeFp64(self._max_capacity)
+            self._min_tree = MinSegmentTreeFp64(self._max_capacity)
+        else:
+            raise NotImplementedError(
+                f"dtype {dtype} not supported by PrioritizedReplayBuffer"
+            )
+        self._max_priority = 1.0
+
+    @property
+    def default_priority(self) -> float:
+        return (self._max_priority + self._eps) ** self._alpha
+
+    def sample(self, storage: Storage, batch_size: int) -> torch.Tensor:
+        p_sum = self._sum_tree.query(0, len(storage))
+        p_min = self._min_tree.query(0, len(storage))
+        if p_sum <= 0:
+            raise RuntimeError("negative p_sum")
+        if p_min <= 0:
+            raise RuntimeError("negative p_min")
+        mass = np.random.uniform(0.0, p_sum, size=batch_size)
+        index = self._sum_tree.scan_lower_bound(mass)
+        if isinstance(index, torch.Tensor):
+            index.clamp_max_(len(storage) - 1)
+        else:
+            index = np.clip(index, None, len(storage) - 1)
+        weight = self._sum_tree[index]
+
+        # Importance sampling weight formula:
+        #   w_i = (p_i / sum(p) * N) ^ (-beta)
+        #   weight_i = w_i / max(w)
+        #   weight_i = (p_i / sum(p) * N) ^ (-beta) /
+        #       ((min(p) / sum(p) * N) ^ (-beta))
+        #   weight_i = ((p_i / sum(p) * N) / (min(p) / sum(p) * N)) ^ (-beta)
+        #   weight_i = (p_i / min(p)) ^ (-beta)
+        # weight = np.power(weight / (p_min + self._eps), -self._beta)
+        weight = np.power(weight / p_min, -self._beta)
```

Comment:
For the weight part, we can consider to use the batch min which is weight.min() to replace the global min from the ReplayBuffer. Actually it seems the original paper didn't clarify it clearly and there are two implementations.
1. OpenAI baselines uses the global min which is same as we have now. https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/deepq/replay_buffer.py#L158
2. DeepMind Acme uses the batch min https://github.com/deepmind/acme/blob/5fac34092330fbe7d758adc487c4b97c9f58f777/acme/agents/jax/dqn/losses.py#L71

Since DeepMind uses the batch min/max, we can consider to use that which can remove the _min_tree to slightly improve the performance.

I already made this change in RLMeta https://github.com/facebookresearch/rlmeta/blob/75029c8c1b08f57f19672749efb54f571d9acab8/rlmeta/core/replay_buffer.py#L242

I also have some improvement in the implementation of SegmentTree and will create another PR to merge them in torchrl.

Reply:
@xiaomengy 

> Since DeepMind uses the batch min/max, we can consider to use that which can remove the _min_tree to slightly improve the performance.

got it! Improve performance = runtime, not training performance right?

Reply:
> I also have some improvement in the implementation of SegmentTree and will create another PR to merge them in torchrl

lovely can't wait to see this